python/TestHarness/testers/SchemaDiff.py

```diff
@@ -0,0 +1,192 @@
+                if gold_dict[1] != test_dict[1]:
+                    output += "Schema Files of Different Types: \nTest: " + test +"\nGold: " + gold
+                    self.setStatus(self.fail, 'INVALID SCHEMA(S) PROVIDED')
+                    break
+                # Perform the diff.
+                diff = self.do_deepdiff(gold_dict[0], test_dict[0], specs['rel_err'], specs['abs_zero'], specs['ignored_items'])
+                if diff:
+                    output += "Schema difference detected.\nFile 1: " + gold + "\nFile 2: " + test + "\nErrors:\n"
+                    output += diff
+                    self.setStatus(self.diff, 'SCHEMADIFF')
+                    break
+        return output
+
+    #Many xml files in MOOSE don't actually store their data as a proper array, but as a giant string.
+    #This should fix the giant strings as lists split by whitespace, in-place in the dict.
+    def fix_xml_arrays(self, data):
+        for k, v in data.copy().items():
+            if isinstance(v, dict):
+                data[k] = self.fix_xml_arrays(v)
+            elif isinstance(v, list):
+                data[k] = [self.fix_xml_arrays(i) for i in v]
+            elif k == '#text':
+                del data[k]
+                data[k] = v.split()
+        return data
+
+    def import_xml(self,filepath):
+        import xmltodict
+        with open(filepath,"r") as f:
+            xml = xmltodict.parse(f.read())
+            return self.fix_xml_arrays(xml)
+
+    def import_json(self,filepath):
+        with open(filepath,"r") as f:
+            return json.loads(f.read())
+
+    def do_deepdiff(self,orig, comp, rel_err, abs_zero, exclude_values:list=None):
+        import deepdiff
+        from deepdiff.operator import BaseOperator
+        class testcompare(BaseOperator):
+            def __init__(self, rel_err,abs_zero,types,regex_paths=None):
+                self.rel_err = rel_err
+                self.abs_zero = abs_zero
+                #next two members are necessary for deepdiff constructor to work
+                self.regex_paths = regex_paths
+                self.types = types
+
+            def give_up_diffing(self,level, diff_instance):
+                try:
+                    if level.t1 != level.t2:
+                        x = float(level.t1)
+                        y = float(level.t2)
+
+                        if x < self.abs_zero and y < self.abs_zero:
+                            return True
+                        if self.rel_err == 0 or y == 0:
+                            if abs(x-y) > self.rel_err:
+                                return False
+                        elif abs(x-y)/y > self.rel_err:
+                            return False
+                    return True #if the two items are the same, you can stop evaluating them.
+
+                #Often in XML data is not stored correctly as a list/array, and are instead big strings. This should be fixed with "fix_XML_arrays",
```

Comment:
Should we not put this in `XMLDiff` instead?

Reply:
Isn't possible, its an internal class to the differ

Reply:
Actually maybe it owuld. but we wouldn't want to.

Reply:
The point is that its here to catch these weird pseudo-lists that slip through, in fact it would be least useful in xmldiff